Updating static analysis (#8623)
fishtown-analytics/docs.getdbt.com: website/docs/docs/fusion/new-concepts.md

> ## Rendering strategies

Reviewer: I really think this whole rendering strategy stuff has to be nuked. @joellabes agrees

<Lightbox src="/img/fusion/annotated_steps.png" title="Each dot represents a step in that model's execution (render, analyze, run). The numbers reflect step order across the DAG." alignment="left" width="600px"/>

<Expandable alt_header="Rendering and execution (dbt Core)" is_open="true">
  <video src="/img/fusion/CoreJitRun.mp4" autoPlay loop muted style={{ width: "100%", maxWidth: 950 }} />
</Expandable>

<Constant name="core" /> renders a model, runs it in the warehouse, then moves on to the next model.

<Expandable alt_header="Rendering, analysis, and execution (dbt Fusion engine)" is_open="true">
  <video src="/img/fusion/FusionAotRun.mp4" autoPlay loop muted style={{ width: "100%", maxWidth: 950 }} />
</Expandable>

When configured with `static_analysis: strict`, the <Constant name="fusion_engine" /> renders all models in the project, then produces and statically analyzes every model's logical plan, and only then starts running models in the warehouse.

> By rendering and analyzing all models before execution, and only beginning execution once everything is proven valid, the <Constant name="fusion_engine" /> prevents unnecessary consumption of warehouse resources. The amount of analysis performed is determined by the `static_analysis` setting &mdash; `baseline` or `strict`.

Reviewer 2: Suggest making this active voice and writing out the emdash:
```suggestion
By rendering and analyzing all models before execution, and only beginning execution once everything is proven valid, the <Constant name="fusion_engine" /> prevents unnecessary consumption of warehouse resources. dbt determines the amount of analysis performed using the `static_analysis` setting of `baseline` or `strict`.
```

  <video src="/img/fusion/FusionAotRun.mp4" autoPlay loop muted style={{ width: "100%", maxWidth: 950 }} />
</Expandable>

> The <Constant name="fusion_engine" /> _can_ use **Ahead of Time (AOT) rendering and analysis** when configured with `static_analysis: strict`. In strict mode, it renders all models in the project, then produces and statically analyzes every model's logical plan, and only then will it start running models in the warehouse.

Reviewer: @eliasdefaria  how does JIT and AOT come into play with `baseline`?! do users no longer have control via `unsafe`? i'm confused

Author: See [internal slack thread](https://dbt-labs.slack.com/archives/C0ADU24A2QZ/p1771876527533869?thread_ts=1771876327.612339&cid=C0ADU24A2QZ) for context

> By rendering and analyzing all models ahead of time, and only beginning execution once everything is proven to be valid, strict mode avoids consuming any warehouse resources unnecessarily. By contrast, in baseline mode (the default) or when using <Constant name="core" />, SQL errors are flagged during execution.

Reviewer: ```suggestion
By rendering and analyzing all models ahead of time, and only beginning execution once everything is proven to be valid. how much analysis is done is determined by the value of static_analysis: `baseline` or `strict`. the <Constant name="fusion_engine" /> avoids consuming any warehouse resources unnecessarily.

!TODO insert chart!

By contrast, SQL errors in models run by <Constant name="core" />'s engine will only be flagged by the database itself during execution.
```

> By rendering and analyzing all models ahead of time, and only beginning execution once everything is proven to be valid, strict mode avoids consuming any warehouse resources unnecessarily. By contrast, in baseline mode (the default) or when using <Constant name="core" />, SQL errors are flagged during execution.

Author: Will add chart soon
```suggestion
By rendering and analyzing all models ahead of time, and only beginning execution once everything is proven to be valid. The amount of analysis done is determined by the value of` static_analysis:` &mdash; `baseline` or `strict`. The <Constant name="fusion_engine" /> prevents unnecessary consumption of warehouse resources.

By contrast, SQL errors in models run by <Constant name="core" />'s engine will only be flagged by the database itself during execution.
```

The <Constant name="fusion_engine" /> is unique in that it can statically analyze not just a single model in isolation, but every query from one end of your DAG to the other. Even your database can only validate the query in front of it! Concepts like [information flow theory](https://roundup.getdbt.com/i/156064124/beyond-cll-information-flow-theory-and-metadata-propagation) &mdash; although not incorporated into the dbt platform [yet](https://www.getdbt.com/blog/where-we-re-headed-with-the-dbt-fusion-engine) &mdash; rely on stable inputs and the ability to trace columns DAG-wide.

### Baseline mode: A smooth transition from dbt Core

The <Constant name="fusion_engine" /> defaults to `static_analysis: baseline` mode, inspired by similar type-checking and linting tools like [TypeScript's migration approach](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html), [basedpyright's baseline feature](https://docs.basedpyright.com/latest/benefits-over-pyright/baseline/), and [Pydantic's strict/lax modes](https://docs.pydantic.dev/latest/why/#strict-lax).

> The philosophy behind baseline mode:

Reviewer: mention of it somewhere: "gradual typing" 

```suggestion
The philosophy behind the above-mentioned tools and dbt Fusion's baseline mode is:
```

- **Smooth transition**: Provide a familiar first-time experience for users coming from <Constant name="core" />.
- **Incremental opt-in**: Offer a clear pathway to adopt more <Constant name="fusion" /> features over time.
- **Pragmatic validation**: Catch most SQL errors without requiring a complete project overhaul

> Migrating to <Constant name="fusion" /> involves more than moving YAML around. Some scenarios that can make migration more involved include:

Reviewer: ```suggestion
Migrating to <Constant name="fusion" /> can involve more than moving YAML around. Some scenarios that can make migration more involved include:
```

1. **Limited access to sources**: You don't have access to all the sources and models of a large dbt project.
2. **Intricate Jinja workflows**: Your project uses post-hooks and introspection extensively.
3. **Package compatibility**: Your project depends on packages that aren't yet <Constant name="fusion" />-compatible.
4. **Unsupported SQL features**: Your models or sources use advanced data types (`STRUCT`, `ARRAY`, `GEOGRAPHY`) or built-in functions (`AI.PREDICT`, `JSON_FLATTEN`, `st_pointfromgeohash`) not yet supported by the engine.

Baseline mode lets you start using <Constant name="fusion" /> immediately while you address these scenarios incrementally. As you resolve compatibility issues, you can opt specific models or your entire project into `strict` mode for maximum validation guarantees.

> 

Reviewer: what does this mean? what do i get with baseline?

how does baseline differ b/w LSP and CLI?
- CLI warns and LSP errors? (or they might both warn)

Author: Will research this with you and add in a follow up

  <video src="/img/fusion/FusionJitCompileUnsafe.mp4" autoPlay loop muted style={{ width: "100%", maxWidth: 950 }} />
  Note that `model_d` is rendered AOT, since it doesn't use introspection, but it still has to wait for `introspective_model_c` to be analyzed.
</Expandable>

You still gain significant benefits from `baseline` static analysis compared to no static analysis. As you become more familiar with <Constant name="fusion" />, consider rewriting introspective code to make it eligible for AOT rendering and `strict` static analysis.

## Recapping the differences between engines

<Constant name="dbt_core" />:

- Renders all models just-in-time.
- Never runs static analysis.

The <Constant name="fusion_engine" /> (baseline mode &mdash; default):

- Renders all models just-in-time, similar to <Constant name="core" />.
- Statically analyzes all models just-in-time, catching most SQL errors while providing a familiar migration experience.

The <Constant name="fusion_engine" /> (strict mode):

- Renders all models ahead-of-time, unless they use introspective queries.
- Statically analyzes all models ahead-of-time, guaranteeing nothing runs until the entire project is proven valid.

> ## Configuring `static_analysis`

Reviewer: once your dag goes strict->baseline there's no coming back. same for baseline->off (i think!?!)

how does the command line flag affect behavior? does it override all nodes?

if i run `dbtf compile --static-analysis baseline`, does it override all instances of `strict` to be `baseline` now?

Author: Addtionally will grab in a follow up

The <Constant name="fusion_engine" /> is unique in that it can statically analyze not just a single model in isolation, but every query from one end of your DAG to the other. Even your database can only validate the query in front of it! Concepts like [information flow theory](https://roundup.getdbt.com/i/156064124/beyond-cll-information-flow-theory-and-metadata-propagation) &mdash; although not incorporated into the dbt platform [yet](https://www.getdbt.com/blog/where-we-re-headed-with-the-dbt-fusion-engine) &mdash; rely on stable inputs and the ability to trace columns DAG-wide.

> ### Baseline mode: A smooth transition from dbt Core

Reviewer: not keen on replacing that `static_analysis: strict` is still the laudable goal here. we're just adding that baseline is the best out-of-the-box baseline

The philosophy behind baseline mode:

- **Smooth transition**: Provide a familiar first-time experience for users coming from <Constant name="core" />.
- **Incremental opt-in**: Offer a clear pathway to adopt more <Constant name="fusion" /> features over time.
- **Pragmatic validation**: Catch most SQL errors without requiring a complete project overhaul

Migrating to <Constant name="fusion" /> involves more than moving YAML around. Some scenarios that can make migration more involved include:

1. **Limited access to sources**: You don't have access to all the sources and models of a large dbt project.
2. **Intricate Jinja workflows**: Your project uses post-hooks and introspection extensively.
3. **Package compatibility**: Your project depends on packages that aren't yet <Constant name="fusion" />-compatible.

> 4. **Unsupported SQL features**: Your models or sources use advanced data types (`STRUCT`, `ARRAY`, `GEOGRAPHY`) or built-in functions (`AI.PREDICT`, `JSON_FLATTEN`, `st_pointfromgeohash`) not yet supported by the engine.

Reviewer 2: ```suggestion
4. **Unsupported SQL features**: Your models or sources use advanced data types (`STRUCT`, `ARRAY`, `GEOGRAPHY`) or built-in functions (`AI.PREDICT`, `JSON_FLATTEN`, `st_pointfromgeohash`) not yet supported by the <Constant name="fusion_engine" /> .
```

> Baseline mode lets you start using <Constant name="fusion" /> immediately while you address these scenarios incrementally. As you resolve compatibility issues, you can opt specific models or your entire project into `strict` mode for maximum validation guarantees.

Reviewer 2: ```suggestion
Setting `static_analysis` to `baseline` mode lets you start using <Constant name="fusion" /> immediately while you address these scenarios incrementally. As you resolve compatibility issues, you can opt specific models or your entire project into `strict` mode for maximum validation guarantees.
```

## Recapping the differences between engines

> <Constant name="dbt_core" />:

Reviewer 2: This is not showing up at all.
```suggestion
<Constant name="core" />:
```

- Renders and runs models one at a time.
- Never runs static analysis.

> The <Constant name="fusion_engine" /> (baseline mode &mdash; default):

Reviewer: :kiss:

- renders all models just-in-time
- never runs static analysis

> The dbt Fusion engine (baseline mode &mdash; default):

Author: ```suggestion
The <Constant name="fusion_engine" /> (baseline mode &mdash; default):
```

> - renders all models just-in-time, similar to <Constant name="core" />

Author: ```suggestion
- Renders all models just-in-time, similar to <Constant name="core" />.
```

> - statically analyzes all models just-in-time, catching most SQL errors while providing a familiar migration experience

Author: ```suggestion
- Statically analyzes all models just-in-time, catching most SQL errors while providing a familiar migration experience.
```

- Then it statically analyzes each model's logical plan in order.
- Finally, it runs each model's rendered SQL. Nothing is persisted to the database until Fusion has validated the entire project.

> ### Introspective model with baseline static analysis

Reviewer: @eliasdefaria again, what do we do about `unsafe` vis a vis AoT and JiT? maybe just nuke this section?

- Then it statically analyzes each model's logical plan in order.
- Finally, it runs each model's rendered SQL. Nothing is persisted to the database until Fusion has validated the entire project.

> ### Introspective model with JIT (baseline) static analysis

Author: ```suggestion
### Introspective model with baseline static analysis
```
